python/simulation.py

```diff
@@ -0,0 +1,783 @@
+    basestring
+except NameError:
+    basestring = str
+
+
+CYLINDRICAL = -2
+
+
+def get_num_args(func):
+    return func.__code__.co_argcount
+
+
+def is_bound(f):
+    return hasattr(f, '__self__') and f.__self__ is not None
+
+
+def py_v3_to_vec(dims, v3):
+    if dims == 1:
+        if v3.x == v3.y == v3.z:
+            return mp.vec(dims, v3.x)
+        else:
+            return mp.vec(dims)
+    elif dims == 2:
+        return mp.vec(v3.x, v3.y)
+    elif dims == 3:
+        return mp.vec(v3.x, v3.y, v3.z)
+    else:
+        raise ValueError("Invalid dimensions in Volume: {}".format(dims))
+
+
+class Pml(object):
+
+    def __init__(self, thickness,
+                 direction=-1,
+                 side=-1,
+                 r_asymptotic=1e-15,
+                 mean_stretch=1.0,
+                 pml_profile=lambda u: u * u):
+
+        self.thickness = thickness
+        self.direction = direction
+        self.side = side
+        self.r_asymptotic = r_asymptotic
+        self.mean_stretch = mean_stretch
+        self.pml_profile = pml_profile
+
+        if direction == -1 and side == -1:
+            self.swigobj = mp.pml(thickness, r_asymptotic, mean_stretch)
+        elif direction == -1:
+            self.swigobj = mp.pml(thickness, side, r_asymptotic, mean_stretch)
+        else:
+            self.swigobj = mp.pml(thickness, direction, side, r_asymptotic, mean_stretch)
+
+    @property
+    def r_asymptotic(self):
+        return self._r_asymptotic
+
+    @r_asymptotic.setter
+    def r_asymptotic(self, val):
+        self._r_asymptotic = check_positive('Pml.r_asymptotic', val)
+
+    @property
+    def mean_stretch(self):
+        return self._mean_stretch
+
+    @mean_stretch.setter
+    def mean_stretch(self, val):
+        if val >= 1:
+            self._mean_stretch = val
+        else:
+            raise ValueError("Pml.mean_stretch must be >= 1. Got {}".format(val))
+
+
+class Absorber(Pml):
+    pass
+
+
+class Symmetry(object):
+
+    def __init__(self, direction, phase=1):
+        self.direction = direction
+        self.phase = complex(phase)
+        self.swigobj = None
+
+
+class Rotate2(Symmetry):
+    pass
+
+
+class Rotate4(Symmetry):
+    pass
+
+
+class Mirror(Symmetry):
+    pass
+
+
+class Identity(Symmetry):
+    pass
+
+
+class Volume(object):
+
+    def __init__(self, center, size=Vector3(), dims=2,):
+        self.center = center
+        self.size = size
+        self.dims = dims
+
+        v1 = center - size.scale(0.5)
+        v2 = center + size.scale(0.5)
+
+        vec1 = py_v3_to_vec(self.dims, v1)
+        vec2 = py_v3_to_vec(self.dims, v2)
+
+        self.swigobj = mp.volume(vec1, vec2)
+
+
+class Harminv(object):
+
+    def __init__(self, sim, c, pt, fcen, df, mxbands=None):
+        self.c = c
+        self.pt = pt
+        self.fcen = fcen
+        self.df = df
+        self.mxbands = mxbands
+        self.harminv_data = []
+        self.harminv_data_dt = 0
+        self.harminv_results = []
```

Comment:
the results shold be in some kind of user-accessible format.  Maybe a `harminv_results` should be called `modes` (for resonant modes) and should be in some kind of `Mode` object so that you can do `modes[1].Q` etcetera to access the different quantities.

Reply:
Done.